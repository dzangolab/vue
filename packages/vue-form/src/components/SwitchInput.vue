<template>
  <div :class="`field switch-toggle ${name}`">
    <label v-if="label" :for="name">
      {{ label }}
    </label>
    <Field
      v-slot="{ field, meta }"
      :name="name"
      :rules="fieldSchema"
      type="checkbox"
      @change="onChange"
    >
      <Switch
        v-bind="field"
        :id="`input-field-${name}`"
        :class="{
          invalid: meta.touched && !meta.valid,
        }"
        :model-value="modelValue"
        :disabled="disabled"
        tabindex="0"
      />
      <ErrorMessage :name="name" />
    </Field>
  </div>
</template>

<script lang="ts">
export default {
  name: "SwitchInput",
};
</script>

<script setup lang="ts">
import { toFieldValidator } from "@vee-validate/zod";
import { ErrorMessage, Field } from "vee-validate";
import { z } from "zod";

import Switch from "./Switch.vue";

import type { PropType } from "vue";

const props = defineProps({
  disabled: {
    default: false,
    type: Boolean,
  },
  label: {
    default: "",
    required: false,
    type: String as PropType<string>,
  },
  modelValue: {
    default: false,
    type: Boolean,
  },
  name: {
    default: "switch",
    required: false,
    type: String as PropType<string>,
  },
  schema: {
    default: () => {
      return {};
    },
    required: false,
    type: Object as PropType<z.ZodType<string | number | boolean>>,
  },
});

const emit = defineEmits(["update:modelValue"]);

const fieldSchema = Object.keys(props.schema).length
  ? toFieldValidator(props.schema)
  : null;

const onChange = (event: Event) => {
  const value = (event.target as HTMLInputElement).checked;

  emit("update:modelValue", Boolean(value));
};
</script>

<style lang="css">
.switch-toggle {
  --_switch-field-direction: var(--switch-field-direction, row);
  --_switch-field-gap: var(--form-field-gap, 0.75rem);

  display: flex;
  flex-direction: var(--_switch-field-direction);
  gap: var(--_switch-field-gap);
  width: max-content;
}
</style>
