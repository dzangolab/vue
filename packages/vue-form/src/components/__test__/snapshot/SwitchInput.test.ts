import { mount } from "@vue/test-utils";
import { describe, expect, it } from "vitest";

import SwitchInput from "../../SwitchInput.vue";

describe("SwitchInput", () => {
  it("matches snapshot", () => {
    const wrapper = mount(SwitchInput, {
      props: {
        label: "Switch input",
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
