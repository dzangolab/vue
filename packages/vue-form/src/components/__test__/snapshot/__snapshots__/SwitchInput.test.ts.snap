// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SwitchInput > matches snapshot 1`] = `
<div
  class="field switch-toggle switch"
>
  <label
    for="switch"
  >
    Switch input
  </label>
  
  <div
    checked="true"
    class="switch"
    id="input-field-switch"
    name="switch"
    tabindex="0"
  >
    <!--v-if-->
    <input
      class="switch-input"
      type="checkbox"
    />
  </div>
  <!---->
  
</div>
`;
